twine/cli.py

```diff
@@ -23,15 +22,15 @@
 
 import twine
 from twine import _installed
+from twine import importlib_metadata
 
 args = argparse.Namespace()
 
 
 def _registered_commands(
     group: str = "twine.registered_commands",
-) -> Dict[str, pkg_resources.EntryPoint]:
-    registered_commands = pkg_resources.iter_entry_points(group=group)
-    return {c.name: c for c in registered_commands}
+) -> Dict[str, importlib_metadata.EntryPoint]:
+    return dict(importlib_metadata.entry_points()[group])
```

Comment:
This is failing type checking on 3.8 (and probably 3.9), where Twine is using standard `importlib.metadata` instead of the third-party `importlib_metadata`.

```
twine/cli.py:33: error: Argument 1 to "dict" has incompatible type "Tuple[EntryPoint, ...]"; expected "Iterable[Tuple[str, EntryPoint]]"
```

https://github.com/pypa/twine/pull/728/checks?check_run_id=1737946626#step:5:12

I'm guessing it's not failing w/ `importlib_metadata` because we ignore that missing import:

https://github.com/pypa/twine/blob/87846e5777b380d4704704a69e1f9a7a1231451c/mypy.ini#L25-L27

Reply:
I noticed that too, and I'm not confident that I understand what the problem is. I suspect the issue is that mypy doesn't recognize that an object can act as a tuple when passed to a dict constructor, as [illustrated here](https://github.com/python/importlib_metadata/blob/9950845a8b3f35843057d3708ed75ef30dd62659/importlib_metadata/__init__.py#L143-L145). My temptation is just to ignore the type error and move on.

```suggestion
    return dict(importlib_metadata.entry_points()[group])  # type: ignore
```

Reply:
What about a `cast` instead? Maybe with a comment? Without looking at the docs/source for `importlib.metadata`, it's not obvious what the data structure should be.

Or, could you use a dict comprehension, as in the existing code?

Reply:
Well, the whole point of an `EntryPoint` providing an `__iter__` was to make it simple for downstream consumers to construct a map of EntryPoint objects by name in one line. If every downstream consumer needs to have comments or casts or their own comprehension, then the approach is failing to achieve its objective. My hope was to make this most common usage pattern very simple and concise even though the technique that enables it is somewhat obscure. In other words, I wanted to find a way that if almost every caller wants a dict of the EntryPoints by name, they can do that with a simple construction to dict.

I've filed python/mypy#9938 to report the underlying issue. My guess is it will be closed as a Python usage that should be discouraged. What do you think about this: let's ignore the error for now until that mypy issue is addressed, and if the mypy issue indicates that this usage is effectively invalid Python (even though it's valid and unambiguous in an untyped world), then I'll convert it to a comprehension.

Reply:
After more consideration, I've decided the `dict` constructor hack is both non-intuitive and not fully usable, so I'm preparing an improved interface in python/importlib_metadata#278. In the meantime, I recommend to go with a simple dict comprehension.
```suggestion
    # todo: with python/importlib_metadata#278:
    # return importlib_metadata.entry_points()[group]
    return {ep.name: ep for ep in importlib_metadata.entry_points()[group]}
```

Reply:
Thanks for diving into the weeds on this. Seems like this could be a 👍  or 👎  for type-checking, depending on one's perspective. 😉